File: bot-shared-library/src/main/java/gg/mineral/bot/impl/controls/Mouse.java
package gg.mineral.bot.impl.controls;

import java.util.Iterator;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

import gg.mineral.bot.api.controls.MouseButton.Type;
import gg.mineral.bot.api.debug.Logger;
import gg.mineral.bot.api.event.EventHandler;
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent;
import gg.mineral.bot.api.instance.ClientInstance;
import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
import lombok.Getter;
import lombok.Setter;
import lombok.val;

public class Mouse implements gg.mineral.bot.api.controls.Mouse, Logger {

    private final MouseButton[] mouseButtons;
    @Getter
    @Setter
    private int dWheel;
    @Getter
    private int x, y;
    @Setter
    private int dX, dY;
    private final Queue<Log> logs = new ConcurrentLinkedQueue<>();
    private Log eventLog = null, currentLog = null;
    private Iterator<Log> iterator = null;
    private final EventHandler eventHandler;

    private final Object2LongOpenHashMap<Runnable> scheduledTasks = new Object2LongOpenHashMap<>();

    public Mouse(EventHandler eventHandler) {
        this.eventHandler = eventHandler;
        mouseButtons = new MouseButton[MouseButton.Type.values().length];

        for (val type : MouseButton.Type.values())
            mouseButtons[type.ordinal()] = new MouseButton(type);
    }

    public void onGameLoop(long time) {
        scheduledTasks.keySet().removeIf(runnable -> {
            if (time >= scheduledTasks.getLong(runnable)) {
                runnable.run();
                return true;
            }
            return false;
        });
    }

    public void schedule(Runnable runnable, long delay) {
        scheduledTasks.put(runnable, (System.nanoTime() / 1000000) + delay);
    }

    @Override
    public MouseButton getButton(MouseButton.Type type) {
        return mouseButtons[type.ordinal()];
    }

    @Override
    public void pressButton(int durationMillis, MouseButton.Type... types) {
        for (val type : types) {
            val button = getButton(type);

            if (button == null || button.isPressed())
                return;

            val event = new MouseButtonEvent(type, true);

            if (eventHandler.callEvent(event))
                return;

            info("Pressing button: " + type);
            button.setPressed(true);
            if (currentLog != null)
                logs.add(currentLog);
            currentLog = new Log(type, true, x, y, dX, dY,
                    dWheel);
            if (durationMillis > 0 && durationMillis < Integer.MAX_VALUE)
                schedule(() -> unpressButton(type), durationMillis);
        }
    }

    @Override
    public void unpressButton(int durationMillis, MouseButton.Type... types) {
        for (val type : types) {
            val button = getButton(type);

            if (button == null || !button.isPressed())
                return;

            val event = new MouseButtonEvent(type, false);

            if (eventHandler.callEvent(event))
                return;

            info("Unpressing button: " + type);
            button.setPressed(false);
            if (currentLog != null)
                logs.add(currentLog);
            currentLog = new Log(type, false, x, y, dX, dY, dWheel);
        }
    }

    @Override
    public boolean next() {
        if (currentLog != null) {
            this.logs.add(currentLog);
            currentLog = null;
        }

        if (iterator == null) {
            buttonsLoop: for (val button : mouseButtons) {
                if (button.isPressed()) {
                    for (val log : logs)
                        if (log.type == button.getType())
                            continue buttonsLoop;

                    logs.add(new Log(button.getType(), true, x, y, dX, dY, dWheel));
                }
            }

            iterator = logs.iterator();
        }

        if (iterator.hasNext()) {
            eventLog = iterator.next();
            iterator.remove();
            return true;
        }

        iterator = null;

        return false;
    }

    @Override
    public MouseButton.Type getEventButtonType() {
        return eventLog != null ? eventLog.type : null;
    }

    public boolean isButtonDown(int i) {
        val type = MouseButton.Type.fromKeyCode(i);
        if (type == null)
            return false;
        return isButtonDown(type);
    }

    public boolean isButtonDown(Type type) {
        return getButton(type).isPressed();
    }

    @Override
    public int getEventButton() {
        return eventLog != null ? eventLog.type.getKeyCode() : -1;
    }

    public record Log(MouseButton.Type type, boolean pressed, int x, int y, int dX, int dY,
            int dWheel) {
    }

    @Override
    public void setX(int x) {
        this.dX = x - this.x;
        this.x = x;
        if (currentLog != null)
            currentLog = new Log(currentLog.type, currentLog.pressed, x, y, dX, dY, dWheel);
        else
            currentLog = new Log(Type.UNKNOWN, false, x, y, dX, dY, dWheel);
    }

    @Override
    public void setY(int y) {
        this.dY = y - this.y;
        this.y = y;
        if (currentLog != null)
            currentLog = new Log(currentLog.type, currentLog.pressed, x, y, dX, dY, dWheel);
        else
            currentLog = new Log(Type.UNKNOWN, false, x, y, dX, dY, dWheel);
    }

    @Override
    public int getDX() {
        try {
            return dX;
        } finally {
            dX = 0;
        }
    }

    @Override
    public int getDY() {
        try {
            return dY;
        } finally {
            dY = 0;
        }
    }

    public int getEventDWheel() {
        return eventLog != null ? eventLog.dWheel : 0;
    }

    public int getEventX() {
        return eventLog != null ? eventLog.x : x;
    }

    public int getEventY() {
        return eventLog != null ? eventLog.y : y;
    }

    public boolean getEventButtonState() {
        return eventLog != null && eventLog.pressed;
    }

    @Override
    public void stopAll() {
        unpressButton(MouseButton.Type.values());
        scheduledTasks.clear();
    }

    @Override
    public void changeYaw(float dYaw) {
        val defaultMouseSense = 0.5f;
        val sensitivity = defaultMouseSense * 0.6F + 0.2F;
        val deltaX = dYaw / (sensitivity * sensitivity * sensitivity * 8.0F);
        this.setDX((int) (deltaX / 0.15));
    }

    @Override
    public void changePitch(float dPitch) {
        val defaultMouseSense = 0.5f;
        val sensitivity = defaultMouseSense * 0.6F + 0.2F;
        // TODO: inverted
        // val inverted = this.mc.gameSettings.invertMouse ? -1 : 1;
        val deltaY = -dPitch / (sensitivity * sensitivity * sensitivity * 8.0F /* * inverted */
        );
        this.setDY((int) (deltaY / 0.15));
    }

    @Override
    public boolean isLoggingEnabled() {
        return eventHandler instanceof ClientInstance instance && instance.getConfiguration() != null
                && instance.getConfiguration().isDebug();
    }
}

File: bot-shared-library/src/main/kotlin/gg/mineral/bot/impl/controls/Mouse.kt
package gg.mineral.bot.impl.controls

import gg.mineral.bot.api.event.EventHandler
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent
import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

open class Mouse(private val eventHandler: EventHandler) : gg.mineral.bot.api.controls.Mouse {
    private val mouseButtons: Array<MouseButton> =
        gg.mineral.bot.api.controls.MouseButton.Type.entries.map { MouseButton(it) }.toTypedArray()

    override var dWheel: Int = 0

    override var x: Int = 0
        set(value) {
            this.dX = value - field
            field = value
            currentLog = currentLog?.let {
                Log(
                    it.type,
                    it.pressed,
                    value,
                    y,
                    dX,
                    dY,
                    dWheel
                )
            } ?: Log(
                gg.mineral.bot.api.controls.MouseButton.Type.UNKNOWN,
                false,
                value,
                y,
                dX,
                dY,
                dWheel
            )
        }

    override var y: Int = 0
        set(value) {
            this.dY = value - field
            field = value
            currentLog = currentLog?.let {
                Log(
                    it.type,
                    it.pressed,
                    x,
                    value,
                    dX,
                    dY,
                    dWheel
                )
            } ?: Log(
                gg.mineral.bot.api.controls.MouseButton.Type.UNKNOWN,
                false,
                x,
                value,
                dX,
                dY,
                dWheel
            )
        }

    override var dX: Int = 0
        get() {
            val result = field
            field = 0
            return result
        }

    override var dY: Int = 0
        get() {
            val result = field
            field = 0
            return result
        }
    private val logs: Queue<Log> = ConcurrentLinkedQueue()
    private var eventLog: Log? = null
    private var currentLog: Log? = null
    private var iterator: MutableIterator<Log>? = null
    override var isGrabbed = true

    private val scheduledTasks = Object2LongOpenHashMap<Runnable>()

    fun onGameLoop(time: Long) {
        scheduledTasks.keys.removeIf { runnable: Runnable ->
            if (time >= scheduledTasks.getLong(runnable)) {
                runnable.run()
                return@removeIf true
            }
            false
        }
    }

    private fun schedule(runnable: Runnable, delay: Long) {
        scheduledTasks.put(runnable, (System.nanoTime() / 1000000) + delay)
    }

    override fun getButton(type: gg.mineral.bot.api.controls.MouseButton.Type): MouseButton {
        return mouseButtons[type.ordinal]
    }

    override fun pressButton(durationMillis: Int, vararg types: gg.mineral.bot.api.controls.MouseButton.Type) {
        for (type in types) {
            val button = getButton(type)

            if (button.isPressed) continue

            val event = MouseButtonEvent(type, true)

            if (eventHandler.callEvent(event)) continue

            logger.debug("Pressing button: {}", type)
            button.isPressed = true
            if (currentLog != null) logs.add(currentLog)
            currentLog = Log(
                type, true, x, y, dX, dY,
                dWheel
            )
            if (durationMillis > 0 && durationMillis < Int.MAX_VALUE) schedule(
                { unpressButton(type) },
                durationMillis.toLong()
            )
        }
    }

    override fun unpressButton(durationMillis: Int, vararg types: gg.mineral.bot.api.controls.MouseButton.Type) {
        for (type in types) {
            val button = getButton(type)

            if (!button.isPressed) continue

            val event = MouseButtonEvent(type, false)

            if (eventHandler.callEvent(event)) continue

            logger.debug("Unpressing button: {}", type)
            button.isPressed = false
            if (currentLog != null) logs.add(currentLog)
            currentLog = Log(type, false, x, y, dX, dY, dWheel)
        }
    }

    override fun next(): Boolean {
        if (currentLog != null) {
            logs.add(currentLog)
            currentLog = null
        }

        if (iterator == null) {
            buttonsLoop@ for (button in mouseButtons) {
                if (button.isPressed) {
                    for ((type) in logs) if (type == button.type) continue@buttonsLoop

                    logs.add(Log(button.type, true, x, y, dX, dY, dWheel))
                }
            }

            iterator = logs.iterator()
        }

        if (iterator?.hasNext() == true) {
            eventLog = iterator?.next()
            iterator?.remove()
            return true
        }

        iterator = null

        return false
    }

    override val eventButtonType: gg.mineral.bot.api.controls.MouseButton.Type?
        get() = eventLog?.type

    open fun isButtonDown(i: Int): Boolean {
        val type: gg.mineral.bot.api.controls.MouseButton.Type =
            gg.mineral.bot.api.controls.MouseButton.Type.fromKeyCode(i)
        return isButtonDown(type)
    }

    open fun isButtonDown(type: gg.mineral.bot.api.controls.MouseButton.Type): Boolean {
        return getButton(type).isPressed
    }

    override val eventButton: Int
        get() = eventLog?.type?.keyCode ?: -1

    @JvmRecord
    data class Log(
        val type: gg.mineral.bot.api.controls.MouseButton.Type,
        val pressed: Boolean,
        val x: Int,
        val y: Int,
        val dX: Int,
        val dY: Int,
        val dWheel: Int
    )

    override fun setCursorPosition(x: Int, y: Int) {
        this.x = x
        this.y = y
    }

    open val eventDWheel: Int
        get() = eventLog?.dWheel ?: 0

    val eventX: Int
        get() = eventLog?.x ?: x

    val eventY: Int
        get() = eventLog?.y ?: y

    open val eventButtonState: Boolean
        get() = eventLog?.pressed == true

    override fun stopAll() {
        unpressButton(*gg.mineral.bot.api.controls.MouseButton.Type.entries.toTypedArray())
        scheduledTasks.clear()
    }

    override fun changeYaw(dYaw: Float) {
        val defaultMouseSense = 0.5f
        val sensitivity = defaultMouseSense * 0.6f + 0.2f
        val deltaX = dYaw / (sensitivity * sensitivity * sensitivity * 8.0f)
        this.dX = (deltaX / 0.15).toInt()
    }

    override fun changePitch(dPitch: Float) {
        val defaultMouseSense = 0.5f
        val sensitivity = defaultMouseSense * 0.6f + 0.2f
        // TODO: inverted
        // val inverted = this.mc.gameSettings.invertMouse ? -1 : 1;
        val deltaY = -dPitch / (sensitivity * sensitivity * sensitivity * 8.0f /* * inverted */
                )
        this.dY = (deltaY / 0.15).toInt()
    }

    companion object {
        private val logger: Logger = LogManager.getLogger(Mouse::class.java)
    }
}
